src/ethereum/frontier/vm/instructions/stack.py

```diff
@@ -0,0 +1,175 @@
+"""
+Ethereum Virtual Machine (EVM) stack Instructions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. contents:: Table of Contents
+    :backlinks: none
+    :local:
+
+Introduction
+------------
+
+Implementations of the EVM stack related instructions.
+"""
+
+from functools import partial
+
+from ethereum.base_types import U256
+
+from .. import Evm
+from ..gas import GAS_VERY_LOW, subtract_gas
+from ..stack import push
+
+
+def push_n(evm: Evm, num_bytes: int) -> None:
+    """
+    Pushes a N-byte immediate onto the stack.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+
+    num_bytes :
+        The number of immediate bytes to be read from the code and pushed to
+        the stack.
+
+    Raises
+    ------
+    StackOverflowError
+        If `len(stack)` is equals `1024`.
+    OutOfGasError
+        If `evm.gas_left` is less than `GAS_VERY_LOW`.
+    """
+    assert evm.pc + num_bytes < len(evm.code)
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_VERY_LOW)
+
+    data_to_push = U256.from_be_bytes(
+        evm.code[evm.pc + 1 : evm.pc + num_bytes + 1]
+    )
+    push(evm.stack, data_to_push)
+
+    evm.pc += num_bytes
+
+
+def dup_n(evm: Evm, item_number: int) -> None:
+    """
+    Duplicate the Nth stack item (from top of the stack) to the top of stack.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+
+    item_number :
+        The stack item number (0-indexed from top of stack) to be duplicated
+        to the top of stack.
+
+    Raises
+    ------
+    OutOfGasError
+        If `evm.gas_left` is less than `GAS_VERY_LOW`.
+    """
+    assert item_number < len(evm.stack)
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_VERY_LOW)
+
+    data_to_duplicate = evm.stack[len(evm.stack) - 1 - item_number]
+    push(evm.stack, data_to_duplicate)
+
+
+def swap_n(evm: Evm, item_number: int) -> None:
+    """
+    Swap the 1st and Nth items in the stack. All items are 0-indexed from the
+    top of the stack.
```

Comment:
Is `swap_n(evm, 0)` a no-op? Either way, it might be nice to show what it does in the docstring.

Reply:
swap_n with n = 0 is a no-op but it isn't a valid opcode. So do we still mention about it?

Reply:
Yeah, I think either a note or an assert would be nice to have.

Reply:
Cool will mention in the docstring.